### ISSUE TITLE
Include responder in middleware process_resource

### ISSUE THREAD

**Issue by timc13** [falconry/falcon#1629]
I like the flexibility that responder name suffixes offers - it would be great if this flexibility were extended to the middleware.  For example, when writing middleware, you cannot distinguish between a non-suffixed responder and a suffixed responder for any given HTTP method.

A nice API for this might be to pass down the actual responder callable to the `process_resource` hook

**Comment by vytas7** [comment]
Hi @timc13!
And thank you for this design proposal, I have labeled the issue accordingly.

For the time being, note that the request object already exposes the [`req.uri_template`](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Request.uri_template) attribute which could probably in some cases be useful for deriving the _suffix_.

A couple of alternative ideas off the top of my head:
* Add a new request attribute telling which _suffix_ was matched, if any (not sure if it is worth the overhead?).
* Add a `CompiledRouter` method to get the _suffix_ corresponding to the given _uri_template_, if any. One could then pass the [`req.uri_template`](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Request.uri_template) to that method to obtain the _suffix_.

**Comment by timc13** [comment]
Those seem like fine workarounds - although maybe a bit indirect.  As far as i'm concerned, it's not the suffix that is important or need be exposed, it's the actual responder method itself and perhaps the method name.  As a counterpoint, I think [this proposal - #709](https://github.com/falconry/falcon/issues/709) is also a good idea.  If a custom `get_responder` is ever used, suffixes may be irrelevant. 

My ultimate wish, is to be able to decorate responder methods and attach some metadata to them at declaration time.  Then my middleware could `process_resource` with that metadata.

Is there a trade-off to including the exposing the `responder` to the middleware that I'm not seeing?

**Comment by vytas7** [comment]
> Is there a trade-off to exposing the responder to the middleware that I'm not seeing?
Just adding more parameters to the middleware interface is a bit off trade-off in itself; both from the compatibility point of view (we'll have to provide shims), and well as just keeping the arity reasonably low.
But by that I don't mean that I dislike your proposal. I just wanted to provide some alternatives so we can compare and contrast :slightly_frowning_face:

The mentioned [idea from #709](https://github.com/falconry/falcon/issues/709#issuecomment-190437202) is also a possible way to solve this.

**Comment by vytas7** [comment]
FWIW, this was raised again (on Gitter), see also https://github.com/falconry/falcon/discussions/1818.

There, I was thinking to myself, maybe we could define a `Route` class (could be a `namedtuple` or something similar, rigid and clean) that would contain all the relevant info: URI template, responder method, HTTP verb, suffix (if any) etc.
And we would then set `req.route` instead of `req.uri_template` (`req.uri_template` could be left as a compatibility wrapper property).
Not sure if this wouldn't be too much burden to implement for custom routers though? :thinking:

**Comment by CaselIT** [comment]
I like the idea.

A drawback to `namedtuple`s is that they are slow to instanciate and also to access parameters (at least until 3.8).
I'm increasingly replacing them with frozen dataclasses with slots.

**Comment by vytas7** [comment]
Yes, sorry, I merely meant that these objects must be reasonably fast to access, and ideally immutable, but their instantiation may take time since that would happen only upon adding/reshuffling routes.

**Comment by CaselIT** [comment]
> **Edit**: good to know they are slow even to access, sounds like that defeats the purpose a bit.

py 3.8 mostly fixed them, but before it the access to named attributes of a named tuple was basically implemented as a `@property`, so definitely slower then a normal attribute lockup

**Comment by vytas7** [comment]
It still benchmarks slightly slower than a normal attribute on a bare class object, but the difference is rather negligible. Much faster than a normal property.
